ESQL: Enable async get to support formatting

kanoshiou opened the pull request:
An async request in text mode (`POST _query/async?format=txt`) previously returned an empty body without providing an async ID. I have now implemented the second approach mentioned in the issue #110926. Now, the async parameters (`Async-ID` & `Async-running`) are returned as headers.

However, I discovered an issue: when using the async ID from a text mode request to query `/_query/async/{id}`, the returned data format is still JSON. This might be a bug, and perhaps the response should be in plain text format, matching the initial request.

Closes #110926

kanoshiou committed:
Async request in txt mode returns the async parameters as headers

kanoshiou committed:
Add changelog

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

bpintea commented:
Thanks for your interest, @kanoshiou.

We'll want to make sure that all operations are possible and work in `txt` mode with the async querying, not just starting the query (see other [async APIs](https://www.elastic.co/guide/en/elasticsearch/reference/master/esql-apis.html)); since if that's not an option for some reason, we'll want to reject the async "txt query" all together, rather than add the headers.

Besides that any change would require tests. For `txt` mode we'll want (besides possible unit tests) integration tests. Have a look at x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java.

kanoshiou committed:
Typo

kanoshiou committed:
Change listener of `/_query/async/{id}` to `EsqlResponseListener` to support format and delimiter in params

kanoshiou committed:
Change HTTP header names access modifiers to private

kanoshiou committed:
Revert "Change HTTP header names access modifiers to private"

This reverts commit e81487a54b1f60dbedf1f67e098a3aaaaeded0a9.

kanoshiou committed:
Response media type defaults to JSON if there is no specific media type in request

kanoshiou committed:
Add tests for async query in text mode

kanoshiou committed:
Get requests do not need checkNonNullMediaType

kanoshiou committed:
Update doc

kanoshiou commented:
Hi @bpintea , 

I believe I have completed the code. Could you please review it when you have some time?

I've updated the listener for `GET /_query/async/{id}` to `EsqlResponseListener`, so it now accepts parameters (`delimiter`, `drop_null_columns` and `format`) like the `POST /_query` API. Additionally, I have added tests to verify the correctness of the code.

You can now set the `format` in the request parameters to specify the return style. I’m not sure if this meets your requirements. If there are any other issues with the code, please let me know, and I will continue to modify it.

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
Merge remote-tracking branch 'origin/main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou commented:
Hi @bpintea ,

I hope this message finds you well. I wanted to gently remind you about https://github.com/elastic/elasticsearch/pull/111104. It has been a while since the submission, and I wanted to check if there are any updates or if any further actions are required from my side.

Thank you for your time and assistance!

bpintea commented:
@kanoshiou sorry for getting back late.
Meanwhile, some of the issues covered in your PR had already been addressed, like the headers insertion, now also aligned with the `_search` API (see #111840).
However, not all of it, specifically the async get, which still doesn't support formatting, neither the originally asked for one, nor one requested as URL parameters (the `format` one). Worst, it doesn't even consider the Accept header formats. Your PR does address this and would be worth continuing.

bpintea commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -342,21 +345,33 @@ public void testTextMode() throws IOException {
         bulkLoadTestData(count);
         var builder = requestObjectBuilder().query(fromIndex() + " | keep keyword, integer | sort integer asc | limit 100");
-        assertEquals(expectedTextBody("txt", count, null), runEsqlAsTextWithFormat(builder, "txt", null));
+        if (mode == ASYNC) {
+            assertEquals(expectedTextBody("txt", count, null), runEsqlAsyncAsTextWithFormat(builder, "txt", null));
+        } else {
+            assertEquals(expectedTextBody("txt", count, null), runEsqlAsTextWithFormat(builder, "txt", null));
+        }
```
Generally, writing smth like the following is easier to read:
```
assertEquals(expectedTextBody("txt", count, null), mode == ASYNC
    ? runEsqlAsyncAsTextWithFormat(builder, "txt", null)
    : runEsqlAsTextWithFormat(builder, "txt", null));
```

bpintea commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -1078,6 +1093,95 @@ static String runEsqlAsTextWithFormat(RequestObjectBuilder builder, String forma
         return Streams.copyToString(new InputStreamReader(entity.getContent(), StandardCharsets.UTF_8));
     }
 
+    static String runEsqlAsyncAsTextWithFormat(RequestObjectBuilder builder, String format, @Nullable Character delimiter)
```
There's now a `runEsqlSync()` (or even a `runEsql()` which takes a mode), it'd be worth refactoring that go get out a result.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -103,24 +110,23 @@ public EsqlResponseListener(RestChannel channel, RestRequest restRequest, EsqlQu
+
+        this.channel = channel;
+        this.restRequest = restRequest;
+        this.esqlQuery = null;
```
The `esqlQuery` is used for logging -- an altnative to making this `null` is to have an `esqlQueryOrId` that could also store the async ID it's being asked for.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlMediaTypeParser.java:
```diff
@@ -43,6 +43,23 @@ public static MediaType getResponseMediaType(RestRequest request, EsqlQueryReque
+            ParsedMediaType acceptType = request.getParsedAccept();
+            mediaType = acceptType != null ? acceptType.toMediaType(MEDIA_TYPE_REGISTRY) : request.getXContentType();
+        }
+        if (mediaType == null) {
```
This should not happen. Check the comments on top of the other `getResponseMediaType()` method.

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
small changes

kanoshiou committed:
Refactor `runEsqlAsync` to support text format

kanoshiou committed:
Refactor `runEsqlSync` to support text format

kanoshiou commented:
@bpintea 

I have updated the code based on your comments. Additionally, I have modified `runEsqlSync` and `runEsqlASync` to support text format. However, I feel a bit uneasy about this part of the code for some reason.

I have one last question: should I rearrange the code before committing? It seems that doing so might change the positions of the methods, making the diff less noticeable.

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
Overload `runEsql` to support text formats

bpintea commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -1007,34 +1020,77 @@ public static Map<String, Object> runEsqlAsync(RequestObjectBuilder requestObjec
+        boolean isTextFormat = format != null;
         if (randomBoolean()) {
-            options.addHeader("Accept", mediaType);
+            if (isTextFormat) {
+                switch (format) {
+                    case "txt" -> options.addHeader("Accept", "text/plain");
+                    case "csv" -> options.addHeader("Accept", "text/csv");
+                    case "tsv" -> options.addHeader("Accept", "text/tab-separated-values");
+                }
+            } else {
+                options.addHeader("Accept", mediaType);
+            }
```
If this needs to be duplicated, build a static map of text formats to Accept values and then you can compact this (and elsewhere) to smth like:
```
options.addHeader("Accept", TEXT_FORMATS.getOrDefault(mediaType, mediaType));
```
And similar in the alternate branch.

However, it might be clearer if you made `runEsqlAsTextWithFormat()` take a format and add implementation for async (potentially still refactoring, to reuse code, where it makes sense), rather than carry on the format&delimiter around and then wrap the textual results into a map just to conform to the return type of the existing non-text methods.

bpintea reviewed (commented):
Left some comments, most importantly the headers must no longer be added.

bpintea commented:
One more note, do a precommit check too please, as noted [here](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#running-verification-tasks).

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
Replace async headers

kanoshiou committed:
Revert

kanoshiou committed:
Add a private constructor for `EsqlResponseListener`

kanoshiou committed:
Check non-null media type

kanoshiou committed:
Refactor `runEsqlAsTextWithFormat` to support async mode

kanoshiou commented:
@bpintea Thank you again for reviewing my PR and providing such valuable suggestions. I have made the necessary changes to the code based on your comments.

Your guidance has been incredibly helpful. I appreciate your patience and support.

kanoshiou committed:
Remove static map `TEXT_FORMATS`

kanoshiou committed:
Update 111104.yaml

kanoshiou committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

kanoshiou committed:
Update 111104.yaml

kanoshiou committed:
Update 111104.yaml

bpintea commented:
@elasticsearchmachine test this please

bpintea reviewed (commented):
Just some minor notes, looking good otherwise.

kanoshiou committed:
Update

kanoshiou commented:
@bpintea Thank you very much! I have updated the branch based on your comments.

bpintea committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

bpintea commented:
@elasticsearchmachine test this please

kanoshiou commented:
Hi @bpintea, I would like to know how I should resolve these test failures, as I can't see the details.

bpintea committed:
Allow an ASYNC GET with no media type spec

Default to JSON in this case.

bpintea committed:
Merge branch 'main' into fork/kanoshiou/async-parameters-returns-as-headers-in-text-mode

bpintea commented:
> I would like to know how I should resolve these test failures

I'm not sure how/if you can access the test results. But most of them will be revealed if you run locally `./gradlew -p x-pack/plugin/esql/ check`. That would take a while, though since it runs BWC tests as well. You can probably run individual tests too (see `./gradlew tasks --all`).
Besides that, another failed test was `./gradlew ":test:external-modules:test-esql-heap-attack:javaRestTest"`.

The tests failed because of missing media type indication. We're a bit inconsistent here, but ES allows a (search) GET with no media type indication whatsoever, so I need to revise my previous comment about supporting a default -- I've pushed a fix for that. However, I'd kindly ask you to add tests for this (i.e. an async GET with no media indication) then too, both `RestEsqlTestCase` and `EsqlMediaTypeParserTests`. Thanks.

bpintea commented:
@elasticsearchmachine test this please

kanoshiou commented:
@bpintea Thank you for letting me know.

> I'd kindly ask you to add tests for this (i.e. an async GET with no media indication) then too, both `RestEsqlTestCase` and `EsqlMediaTypeParserTests`. Thanks.

 Sure thing! I will implement the tests ASAP.

kanoshiou committed:
Testcases for request with no Content-type header

kanoshiou commented:
@bpintea I've added the tests. Please review.

bpintea commented:
@elasticmachine ok to test

bpintea commented:
@elasticsearchmachine test this please

bpintea committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

bpintea commented:
@elasticsearchmachine test this please

bpintea committed:
Merge branch 'main' into async-parameters-returns-as-headers-in-text-mode

bpintea commented:
@elasticsearchmachine test this please

bpintea reviewed (commented):
Looks good, thanks @kanoshiou.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/118257)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/118257">](https://github.com/elastic/elasticsearch/pull/118257) |

kanoshiou commented:
@bpintea, Thank you so much for taking the time to review my code. I truly appreciate your patience and attention to detail. This has been an unforgettable experience, and I have learned a lot.😊
